### DISCUSSION TITLE
2023.05.26

### DISCUSSION THREAD

**Discussion by mywnajsldkf** [Blue-club/pintos_5#2] [개발 일지]
# 오늘 목표
최광민
- [ ] 카이스트 pintos 번역본 읽기
- [ ] 내장된 코드 이해하고 주석 작성하기

장혁
- [ ] 카이스트 pintos 번역본 읽기
- [ ] 내장된 코드 이해하고 주석 작성하기

김정인
- [x] pintos gitbook 읽기
- [ ] 개발 시작해보기

**Comment by chwangmin** [comment]
# Alarm clock

- ‘프로세스를 일정 시간이 지나면 다시 깨우는 기능’

![KakaoTalk_20230527_083628887](https://github.com/Blue-club/pintos_5/assets/44563065/4b8f98bb-a5ef-4e83-b653-15482bb2e068)



### 현재 코드(기본)

- ready state 와 running state사이를 `buzy wait` 상태로 계속 반복하고 있다.
- ex) 5분뒤에 깨워줘~ 하고 1초마다 5분 지났냐 물어보는 꼴…..

### 구현할 코드

- 이러한 비효율을 해결하기 위해 잠이 든 스레드를 ready_state가 아닌, `block state`로 두어서 깨어날 시간이 되기 전까지 스케줄링에 포함되지 않도록 하고, 깨어날 시간이 되었을 때 ready state로 바꾸어 주면 된다.
- sleep_list 을 init 해준다.
- 같은 시간에 일어날 사람이 두명이 있다면? (sort되 있어 1사람만 깨우는 꼴이 될 수 있음) 고려.

### 코드 해석

- timer_elapsed(start)는 timer_sleep이 호출된 시점에서 몇 tick이 지났는지를 반환하는 함수이다.
- 

# 인터럽트 사용 이유

- CPU가 특정 기능을 수행하는 도중에 `급하게 다른 일을 처리`하고자 할 때 사용할 수 있는 기능
- `우선 순위가 급한 일을 처리`할 필요가 있을 때 대처할 수 있는 방안이 필요하다.
- ex) 키보드 키를 누르면 키보드 버퍼에 입력된 후 CPU에 인터럽트가 걸린다. 그럼 현재 처리하던 작업에 대한 정보를 수집하여 저장한 뒤에 인터럽트 서비스 루틴을 수행한다. 인터럽트 처리를 마친 후에는 다시 이전에 처리하던 작업으로 돌아간다.

## 타이머 인터럽트 | 선점형 스케줄러를 위해 필요


- 타이머 인터럽트를 발생시키는 장치가 컴퓨터 안에 칩으로 존재하는데 해당 칩에서 `일정간격으로 인터럽트를 계속해서 발생`시킵니다,

**Comment by mywnajsldkf** [comment]
> **오늘 한 일**
> - gitbook thread 읽기
> - 조각지식 채우기(프로세스, 프로세서, 스레드 / context / scheduling / busy waiting / lock 상태 )
https://blogger.pe.kr/422

스레드 과제 gitbook 정리 내용

![IMG_188FD5345978-1](https://github.com/Blue-club/pintos_5/assets/47661695/729aab0d-349c-4386-acb6-aa991ee3ed06)
![IMG_EE1A9D1587BC-1](https://github.com/Blue-club/pintos_5/assets/47661695/775d925c-1275-4ef3-a938-b5081414f808)
![IMG_0EA40CB33397-1](https://github.com/Blue-club/pintos_5/assets/47661695/47de1500-4043-431a-9649-f06f9c146a11)

> **내일할 일**
> - PintOS 과제 스레드 수업 듣기 https://www.youtube.com/watch?v=myO2bs5LMak
> - 알람 기능 구현
